lumen/ai/llm.py

```diff
@@ -600,4 +604,8 @@ async def invoke(
             if not messages[i]["content"]:
                 messages[i]["content"] = "--"
 
+        if messages and messages[-1]["role"] == "assistant":
+            # cannot end with assistant
```

Comment:
Why not? OpenAI doesn't allow for it?

Reply:
Anthropic doesn't allow for it https://github.com/holoviz/lumen/issues/849

Reply:
Seems hacky though, remind me in which cases we start with and end with assistant?

Starting should only happen if our arbitrary 5(?) message cutoff for context starts with an assistant message? In that case can't we just drop that message or go one further back in history?

And for the ending message can we just move it before the user message?

Reply:
I think we arbitrarily insert assistant messages in coordinator:
```

            if isinstance(subagent, SQLAgent):
                custom_agent = next((a for a in self.agents if isinstance(a, AnalysisAgent)), None)
                if custom_agent:
                    custom_analysis_doc = custom_agent.purpose.replace("Available analyses include:\n", "")
                    custom_message = (
                        f"Avoid doing the same analysis as any of these custom analyses: {custom_analysis_doc} "
                        f"Most likely, you'll just need to do a simple SELECT * FROM {{table}};"
                    )
                    custom_messages.append({"role": "user", "content": custom_message})
            if instruction:
                custom_messages.append({"role": "assistant", "content": instruction})
```

```

            if invalidation_assessment:
                messages.append({"role": "assistant", "content": invalidation_assessment + " so do not choose that table."})
                context_length += 1
```

Reply:
Okay logging it, I realized big issues with how I'm inserting these messages "--"
<img width="1215" alt="image" src="https://github.com/user-attachments/assets/54a55d13-cc90-41d4-a3bd-59377249b60d" />

Reply:
Okay, might add a `system_guidelines`